### PR TITLE
FEATURE: add support for One-Click unsubscribe (RFC 8058)

### PR DESCRIPTION
**Pull request by ZogStriP**
We were missing the "List-Unsubscribe-Post" header in emails we sent to allow Yahoo / GMail and others to automagically show a link to unsubscribe.

Internal ref - t/144713

### CODE REVIEW

File: app/controllers/email_controller.rb [hunk 1]
```diff
@@ -46,10 +46,12 @@ def perform_unsubscribe
...
-    raise Discourse::NotFound unless user
-    topic = Topic.find_by(id: params[:topic_id].to_i) if @topic_id
-    @topic = topic if topic && Guardian.new(nil).can_see?(topic)
+
+    raise Discourse::NotFound unless User.find_by_email(@email)
+
+    if @topic_id = params[:topic_id]
+      topic = Topic.find_by(id: @topic_id)
+      @topic = topic if topic && Guardian.new.can_see?(topic)
+    end
```
Comment (davidtaylorhq) [inline app/controllers/email_controller.rb:55]: Is there any logic change here? Or just an unrelated refactor? 👀

File: app/controllers/email_controller.rb [hunk 2]
```diff
@@ -26,7 +26,7 @@ def unsubscribe
   def perform_unsubscribe
     RateLimiter.new(nil, "unsubscribe_#{request.ip}", 10, 1.minute).performed!
 
-    key = UnsubscribeKey.find_by(key: params[:key])
+    key = UnsubscribeKey.includes(:user).find_by(key: params[:key])
```
Comment (davidtaylorhq) [inline app/controllers/email_controller.rb:29]: Am I right in saying this is just a perf optimization, with no behaviour change?

Comment (ZogStriP) [reply app/controllers/email_controller.rb:29]: Yeah, it technically isn't an improvement because it will do 2 SQL requests no matter which way we do it, but at least, with the "includes" it's explicit and matches the rails way of "side loading"

File: app/controllers/email_controller.rb (hunk 1 above)
Comment (ZogStriP) [reply app/controllers/email_controller.rb:55]: No logic changes, just refactor.

Comment (davidtaylorhq) [reply app/controllers/email_controller.rb:55]: I see, ok. Ideally this would be split out into a separate `DEV:` PR. Otherwise `git blame` will make it seem like it's related to the implementation of RFC8058, which would be very confusing for future travellers.

File: app/controllers/email_controller.rb (hunk 2 above)
Comment (davidtaylorhq) [reply app/controllers/email_controller.rb:29]: Like the other comment, ideally I think this would be moved into a separate `DEV:` PR. It's unrelated to the RFC8058 implementation, which makes review & future `git blame` very confusing.

Comment (ZogStriP) [reply app/controllers/email_controller.rb:29]: @davidtaylorhq just split the changes in two commits

**Review by davidtaylorhq — APPROVED** [review]
